### PR TITLE
Fix shape mismatch in `rnn()` of tensorflow_backend

### PR DESCRIPTION
**Pull request by PeterChe1990**
Fix the shape mismatch in `rnn()` with `not unroll` and `mask is not None` of the Tensorflow backend.

Problem: If the rnn cell has any (recurrent) `states[i]` whose shape is different from that of the `output` (`states[0]`), there will raise an `ValueError` when updating that state.

Reason: This is because the `tiled_mask_t` is not updated correctly for each `states[i]` but instead is simply copied from that of the `output` (`states[0]`).

Solution: Tile the `mask_t` with the correct shape of each `states[i]`. Notice that in a similar situation with `unroll is True`, the `tiled_mask_t` is handled correctly.

### CODE REVIEW

**Comment by fchollet** [comment]
Thanks for the PR! Please add a unit test for this case.

File: tests/keras/backend/reference_operations.py [hunk 1]
```diff
@@ -126,8 +126,13 @@ def pool(x, pool_size, strides, padding, data_format, pool_mode):
 
 
 def rnn(x, w, init, go_backwards=False, mask=None, unroll=False, input_length=None):
-    w_i, w_h, w_o = w
+    # Test these recurrent equations:
+    # h_t = w_o( x_{t} w_i + h_{t-1} w_h + s_{t-1} w_s )
```
Comment (taehoonlee) [inline tests/keras/backend/reference_operations.py]: Thank you for the PR, @PeterChe1990. Are these canonical operations for RNN? Is there any reference article?

Comment (PeterChe1990) [reply tests/keras/backend/reference_operations.py]: This is just an extension of the previous rnn function with an additional hidden state `s`. More complicated states are used in other LSTM/GRU-like RNNs, but the example here is simple to make the test succinct.

File: tests/keras/backend/backend_test.py [hunk 2]
```diff
@@ -599,6 +600,93 @@ def rnn_fn(x_k, h_k):
...
+        x_k = K.variable(x)
+        hs0_k = [K.variable(h0), K.variable(s0)]
+        wi_k = K.variable(wi)
+        wh_k = K.variable(wh)
+        ws_k = K.variable(ws)
+        wxs_k = K.variable(wxs)
+        wss_k = K.variable(wss)
+        mask_k = K.variable(mask)
+
+        def rnn_fn(x_k, hs_k):
```
Comment (fchollet) [inline tests/keras/backend/backend_test.py]: Does it really have to be this complex? This is just a unit test.

why not:

```python
def rnn_fn(x, h):
   return x, [x, K.concatenate([x, x], axis=-1)]
```

Comment (PeterChe1990) [reply tests/keras/backend/backend_test.py]: Thanks for your suggestions! I simplified the unit_test.
